File: client-ios/Sources/Core/Call/WebRtcResilienceConstants.swift
import Foundation

/// Canonical WebRTC resilience constants shared across all Serenada clients.
/// Run `node scripts/check-resilience-constants.mjs` to verify cross-platform parity.
enum WebRtcResilience {

    // MARK: - Signaling

    static let reconnectBackoffBaseMs = 500
    static let reconnectBackoffCapMs = 5_000
    static let connectTimeoutMs = 2_000
    static let pingIntervalMs = 12_000
    static let pongMissThreshold = 2
    static let wsFallbackConsecutiveFailures = 3

    // MARK: - Join

    static let joinPushEndpointWaitMs = 250
    static let joinConnectKickstartMs = 1_200
    static let joinRecoveryMs = 4_000
    static let joinHardTimeoutMs = 15_000

    // MARK: - Peer Connection

    static let offerTimeoutMs = 8_000
    static let iceRestartCooldownMs = 10_000
    static let nonHostFallbackDelayMs = 4_000
    static let nonHostFallbackMaxAttempts = 2
    static let iceCandidateBufferMax = 50

    // MARK: - TURN

    static let turnFetchTimeoutMs = 2_000
    static let turnRefreshTriggerRatio = 0.8

    // MARK: - Snapshot

    static let snapshotPrepareTimeoutMs = 2_000
}

// MARK: - Nanosecond convenience accessors

extension WebRtcResilience {
    static var reconnectBackoffBaseNs: UInt64 { UInt64(reconnectBackoffBaseMs) * 1_000_000 }
    static var reconnectBackoffCapNs: UInt64 { UInt64(reconnectBackoffCapMs) * 1_000_000 }
    static var connectTimeoutNs: UInt64 { UInt64(connectTimeoutMs) * 1_000_000 }
    static var pingIntervalNs: UInt64 { UInt64(pingIntervalMs) * 1_000_000 }
    static var joinPushEndpointWaitNs: UInt64 { UInt64(joinPushEndpointWaitMs) * 1_000_000 }
    static var joinConnectKickstartNs: UInt64 { UInt64(joinConnectKickstartMs) * 1_000_000 }
    static var joinRecoveryNs: UInt64 { UInt64(joinRecoveryMs) * 1_000_000 }
    static var joinHardTimeoutNs: UInt64 { UInt64(joinHardTimeoutMs) * 1_000_000 }
    static var offerTimeoutNs: UInt64 { UInt64(offerTimeoutMs) * 1_000_000 }
    static var iceRestartCooldownNs: UInt64 { UInt64(iceRestartCooldownMs) * 1_000_000 }
    static var nonHostFallbackDelayNs: UInt64 { UInt64(nonHostFallbackDelayMs) * 1_000_000 }
    static var turnFetchTimeoutNs: UInt64 { UInt64(turnFetchTimeoutMs) * 1_000_000 }
    static var snapshotPrepareTimeoutNs: UInt64 { UInt64(snapshotPrepareTimeoutMs) * 1_000_000 }
}

File: client-android/app/src/main/java/app/serenada/android/call/WebRtcResilienceConstants.kt
package app.serenada.android.call

/**
 * Canonical WebRTC resilience constants shared across all Serenada clients.
 * Run `node scripts/check-resilience-constants.mjs` to verify cross-platform parity.
 */
object WebRtcResilienceConstants {

    // ── Signaling ────────────────────────────────────────────────────
    const val RECONNECT_BACKOFF_BASE_MS = 500L
    const val RECONNECT_BACKOFF_CAP_MS = 5_000L
    const val CONNECT_TIMEOUT_MS = 2_000L
    const val PING_INTERVAL_MS = 12_000L
    const val PONG_MISS_THRESHOLD = 2
    const val WS_FALLBACK_CONSECUTIVE_FAILURES = 3

    // ── Join ─────────────────────────────────────────────────────────
    const val JOIN_PUSH_ENDPOINT_WAIT_MS = 250L
    const val JOIN_CONNECT_KICKSTART_MS = 1_200L
    const val JOIN_RECOVERY_MS = 4_000L
    const val JOIN_HARD_TIMEOUT_MS = 15_000L

    // ── Peer Connection ──────────────────────────────────────────────
    const val OFFER_TIMEOUT_MS = 8_000L
    const val ICE_RESTART_COOLDOWN_MS = 10_000L
    const val NON_HOST_FALLBACK_DELAY_MS = 4_000L
    const val NON_HOST_FALLBACK_MAX_ATTEMPTS = 2
    const val ICE_CANDIDATE_BUFFER_MAX = 50

    // ── TURN ─────────────────────────────────────────────────────────
    const val TURN_FETCH_TIMEOUT_MS = 2_000L
    const val TURN_REFRESH_TRIGGER_RATIO = 0.8

    // ── Snapshot ─────────────────────────────────────────────────────
    const val SNAPSHOT_PREPARE_TIMEOUT_MS = 2_000L
}
